Tarea 1

H-Cote opened the pull request:
Primera versión de la tarea 1. Tengo algunas dudas por lo que me falta completar los incisos c) y d) del ejercicio 2 y escribir los códigos como funciones.

H-Cote committed:
Agrego las capturas de pantalla de la tarea0

H-Cote committed:
Puse las capturas de pantalla en un documento de jupyter

H-Cote committed:
Merge branch 'master' of https://github.com/lbenet/2017-2_TSFisComputacional-1 into tarea0

H-Cote committed:
Primera versión de la tarea 1

Falta poner códigos como funciones y completar los incisos c) y d) del ejercicio 2

H-Cote committed:
Nuevas notas del curso

Merge branch 'master' of https://github.com/lbenet/2017-2_TSFisComputacional-1 into tarea_1

lbenet commented:
Algunos comentarios:

- Nota que `c` en el ejercicio es muy pequeña, 10^{-17}, mientras que la `c` en la celda [20] es `Int64` y *muy* grande ( 10^{17} )!

- Mete tus algoritmos en funciones; no sólo es buena práctica, sino que hace fácil verificar el algoritmo para distintos casos. Las funciones deben regresar algo, aunque en ciertos casos pueden simplemente imprimir cosas; haz que devuelva las raíces de la ecuación! (Este comentario de hecho se aplica a otros ejercicios.)

- Si devuelve a veces `Float64` y a veces `Complex{Float64}` entonces no será estable según el tipo.

- En [46], `return` debe aparecer cuando devuelves algo; normalmente al final de la función, en cuyo caso lo puedes omitir.

- ¿Estás contento con el resultado en [27]?

H-Cote committed:
Puse los cógigos como funciones donde era necesario.
En ejercicio 1 modifiqué los resultados de salida para que la función sea
estable según el tipo y documenté la función. Además, cambié el valor de c por el que se pide.
En el ejercicio 2, aún falta completar los incisos c) y d). La función varianza_rec aún no funciona.
En el ejercicio 3, modifiqué el código del método de Simpson para tener una mejor aproximación al valor de la integral.

UriAceves commented:
**Ejercicio 1)**
* Tu función se llama `h`, evita nombres que no describan al menos mínimamente lo que hace una función.
* Tu función no es *type stable*.
* Para `h(1,1,1*10^-17.)` tu resultado (`Las soluciones son reales: x1= 0.0 + 0.0im y x2= -1.0 - 0.0im`) es incorrecto, podrías checar [otras fuentes](https://www.wolframalpha.com/input/?i=x%5E2+%2B+x+%2B+10%5E(-17)+%3D+0).

**Ejercicio 2)**
* Tus funciones imprimen, sería preferible que regresaran valores.
* Revisa la función en donde implementas el algoritmo recursivo.
* No veo respuesta al inciso *d)*.

**Ejercicio 3)**
* De nuevo, intenta que tus funciones regresen valores, en lugar de sólo imprimirlos.

En general se ve bien tu tarea.

lbenet commented:
Una nota sobre tu tarea y los comentarios de Uriel: la función de hecho es type-stable, gracias a que *no* regresa nada excepto información a la pantalla. Si haces que regrese las raices, tal y como está implementada, entonces no será type-stable.

Piensa en lo que hace Julia cuando `2^1` da lo que uno espera, y `2^-1` da un error. Tú puedes hacer algo así para las raíces de la ecuación de segundo grado.

lbenet commented:
En cuanto al segundo ejercicio, me parece que todo está correcto, aunque *puedes* reusar lo que ya implementaste, como la función promedio. Aquí también se aplica lo de antes; haz que las funciones *regresen* algo; es mejor a que simplemente lo impriman.

lbenet commented:
Falta aún implementar la varianza con el otro método...

lbenet commented:
Este mismo [comentario](https://github.com/lbenet/2017-2_TSFisComputacional-1/pull/23#issuecomment-284289951) se aplica al último problema...

H-Cote committed:
Tercera versión de la tarea 1

Las funciones devuelven los resultados en lugar de imprimirlos.

Se completan lo incisos c) y d).

Modificaciones a los códigos de la fórmula general y la varianza para hacerlos más rápidos.

H-Cote committed:
actualización del repositorio
Merge branch 'master' of https://github.com/lbenet/2017-2_TSFisComputacional-1 into tarea_1

lbenet commented:
Respecto a tu último envío:

- El ejercicio 1 te va a dar un error para a=b=1, c=10^7; hay que corregir eso. Y aún falta cómo estabilizar el algoritmo.

- En el ejercicio 2, tu varianza está mal; ver los resultados en [19] y [21], incluyendo el inciso (c). Me sorprende que no te moleste que los resultados sean distintos!

- El resultado del test del ejercicio 3 indica que hay algo mal. Chécalo, porque la integral de x^2 entre 0 y 9 es 243, lo que es *muy* distinto de tu resultado en [23].

lbenet commented:
Algunos comentarios de la última versión:

- En el ejercicio 1 hay una inestabilidad de tipo sutil, que de hecho se muestra como un error. Simplemente, usa el caso a=c=1, y b=2. Este ejercicio está *casi*. Falta considerar el caso a=0, y cómo estabilizar el algoritmo cuando a*c es muy pequeño...

- En el ejercicio 2, el valor obtenido de la varianza es demasiado distinto del que se obtiene con `var`; algún comentario? Probaste la función `varianza_promedio_rec`?

- Falta documentar *todas* las funciones (ejercicios 2 y 3).

Ya casi...

H-Cote committed:
Cuarta versión de la tarea 1

En el ejercicio 1, se hizo una modificación para el caso de una sola raíz y cuando a = 0. Falta el caso cuando el valor de ac es pequeño.

En los ejercicio 2 y 3, se documentaron las funciones.

En el ejercicio 2, en el inciso d), la función varianza_promedio_rec() devuelve tanto el promedio como la varianza de la siguiente forma ([varianza],[promedio]), cuyos valores son cercanos a los obtenidos con las funciones mean() y var()

lbenet commented:
- En cuanto a la pregunta 1, nota lo siguiente:
```julia
julia> BigFloat(1e-17)
1.00000000000000007154242405462192450852805618492324772617063644020163337700069e-17

julia> BigFloat("1e-17")
9.999999999999999999999999999999999999999999999999999999999999999999999999999967e-18
```

- Más allá de esto, la función sigue sin ser numéricamente estable si `a*c` son muy pequeñas (respecto a `b^2`). Prueba lo mismo usando `Float64` y obtendrás 0.0, a pesar de que el resultado muy cercano a 1.0e-17!

- Respecto a la pregunta 2d, corriste `varianza_promedio_rec` dos veces? Me llama la atención que la función sea ~4 veces más lenta que `vari`. Esto puede tener que ver con la última línea del ciclo `for`, que me parece que la puedes sacar de ahí!

lbenet commented:
Acabo de notar que tu función `f_general` **no** es estable según el tipo: a veces da una raíz y a veces dos!

H-Cote committed:
Quinta versión de la tarea 1.
En el ejercicio 1, se cambió la fórmula general habitual por la de fórmula de Citardauq para tener estabilidad de tipo cuando el producto a*c es pequeño.
En el ejercicio 2, en la función varianza_promedio_rec, se sacó la linea del cálculo de la varianza del ciclo for para hacer más rápido los cálculos, sin embargo la función sigue siendo lenta en comparación con las otras.

lbenet commented:
En la pregunta 1 te fuiste de un extremo a otro. No es necesario obtener *ambas* raíces con la modificación de la fórmula general, sólo la problemática! Lee bien [esta liga](https://math.stackexchange.com/questions/866331/numerically-stable-algorithm-for-solving-the-quadratic-equation-when-a-is-very). Una de las raíces la puedes obtener de la fórmula general, y la otra la obtienes a partir de ése cálculo.

Lo demás está bien.

H-Cote committed:
Sexta versión de la tarea 1.
En el ejercicio 1, se utilizaron la fórmula general junto con la fórmula de Citardauq, para calcular las raices. Esto con el fin de evitar errores cuando el producto de a*c es pequeño. Además, se incluyó como resultado de salida un error cuando el valor de a es igual a cero.

lbenet commented:
Listo, queda aceptada! Gracias!
